src/Stack/Upload.hs

```diff
@@ -203,13 +205,14 @@ mkUploader config us = do
         { upload_ = \tarName bytes -> do
             let formData = [partFileRequestBody "package" tarName (RequestBodyLBS bytes)]
             req2 <- formDataBody formData req1
-            let req3 = applyBasicAuth
+            mreq3 <- applyDigestAuth
                     (encodeUtf8 $ hcUsername creds)
                     (encodeUtf8 $ hcPassword creds)
                     req2
+                    manager
             putStr $ "Uploading " ++ tarName ++ "... "
             hFlush stdout
-            withResponse req3 manager $ \res ->
+            withResponse (fromMaybe req2 mreq3) manager $ \res ->
```

Comment:
I don't think it's necessary. If `mreq3` is `Nothing`, then we're going to retry the same request again, and when it (almost) inevitably fails, we'll display the error message from the server to the user.